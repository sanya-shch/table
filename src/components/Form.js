// @flow

import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { setTable } from "../redux/actions/tableActions";
import type { Dispatch } from "../redux/reducers/tableReducer";

export function Form() {
  const [state, setstate] = useState({ m: 0, n: 0, x: 0 });
  const { m, n, x } = state;

  const dispatch: Dispatch = useDispatch<Dispatch>();

  const handleSubmit = (e: SyntheticEvent<HTMLButtonElement>): void => {


    setstate({ m: 0, n: 0, x: 0 });

    dispatch(setTable(m, n, x));
  };

  const handleChange = (e: SyntheticInputEvent<HTMLInputElement>): void => {
    setstate({ ...state, [e.target.name]: parseInt(e.target.value, 10) });
  };

  return (
    <div className="form">
      <label>
        M:
        <input
          data-testid="input_m"
          type="number"
          min={0}
          value={m}
          name="m"
          onChange={handleChange}
        />
      </label>

      <label>
        N:
        <input
          data-testid="input_n"
          type="number"
          min={0}
          value={n}
          name="n"
          onChange={handleChange}
        />
      </label>

      <label>
        X:
        <input
          data-testid="input_x"
          type="number"
          min={0}
          value={x}
          name="x"
          onChange={handleChange}
        />
      </label>

      <button onClick={handleSubmit}>Build table</button>
    </div>
  );
}

export default Form;
