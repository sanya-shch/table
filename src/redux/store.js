// @flow

import { createStore , applyMiddleware  } from "redux";
import reducers from "./reducers";
import type { Store as AppState } from "./reducers/tableReducer";
import { composeWithDevTools } from "redux-devtools-extension";
import type { Store } from "redux";
import type { ActionType } from "./actions/tableActions";

const store: Store<AppState, ActionType> = createStore(
  reducers,
  composeWithDevTools(applyMiddleware())
);

export default store;


