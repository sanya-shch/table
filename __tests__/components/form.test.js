import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { render, fireEvent } from "@testing-library/react";
import Form from "../../src/components/Form";
import reducers from "../../src/redux/reducers/tableReducer";

describe("the Table component", () => {
  it("Form", () => {
    const store = makeTestStore();

    const { getByText, getByLabelText, getByTestId, container } = render(
      <Provider store={store}>{<Form />}</Provider>
    );

    expect(container).toMatchSnapshot();

    expect(getByLabelText("M:").value).toEqual("0");

    expect(getByLabelText("N:").value).toEqual("0");

    expect(getByLabelText("X:").value).toEqual("0");

    fireEvent.change(getByTestId("input_m"), { target: { value: "1000" } });
    expect(
      getByLabelText("M:", {
        selector: "input"
      }).value
    ).toBe("1000");

    fireEvent.change(getByTestId("input_n"), { target: { value: "10" } });
    expect(
      getByLabelText("N:", {
        selector: "input"
      }).value
    ).toBe("10");

    fireEvent.change(getByTestId("input_x"), { target: { value: "30" } });
    expect(
      getByLabelText("X:", {
        selector: "input"
      }).value
    ).toBe("30");

    fireEvent.click(getByText("Build table"));
    expect(store.dispatch).toHaveBeenCalledWith({
      payload: { m: 1000, n: 10, x: 30 },
      type: "SET_TABLE"
    });
  });
});

function makeTestStore() {
  const store = createStore(reducers, {});
  store.dispatch = jest.fn();
  return store;
}
