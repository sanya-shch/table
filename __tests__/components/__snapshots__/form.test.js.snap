// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Table component Form 1`] = `
<div>
  <div
    class="form"
  >
    <label>
      M:
      <input
        data-testid="input_m"
        min="0"
        name="m"
        type="number"
        value="0"
      />
    </label>
    <label>
      N:
      <input
        data-testid="input_n"
        min="0"
        name="n"
        type="number"
        value="0"
      />
    </label>
    <label>
      X:
      <input
        data-testid="input_x"
        min="0"
        name="x"
        type="number"
        value="0"
      />
    </label>
    <button>
      Build table
    </button>
  </div>
</div>
`;
